task/packrat/src/ereport.rs

```diff
@@ -0,0 +1,405 @@
+                    Ok(encoded) => {
+                        data.write_range(
+                            position..position + encoded.len(),
+                            encoded,
+                        )
+                        .map_err(|_| ClientError::WentAway.fail())?;
+
+                        position += encoded.len();
+
+                        ringbuf_entry!(Trace::MetadataEncoded {
+                            len: encoded.len() as u32
+                        });
+                    }
+                    Err(err) => {
+                        // Encoded VPD metadata was too long, or couldn't be
+                        // represented as a CBOR string.
+                        ringbuf_entry!(Trace::MetadataError(err));
+                    }
+                }
+            }
+            // Begin at ENA 0
+            0
+        };
+
+        // End metadata map.
+        data.write_at(position, CBOR_BREAK)
+            .map_err(|_| ClientError::WentAway.fail())?;
+        position += 1;
+
+        let mut reports = 0;
+        // Beginning with the first
+        for r in self.storage.read_from(begin_ena) {
+            if reports >= limit {
+                break;
+            }
+
+            if first_written_ena.is_none() {
+                first_written_ena = Some(r.ena);
+                // Start CBOR list
+                // XXX(eliza): in theory it might be nicer to use
+                // `minicbor::data::Token::BeginArray` here, but it's way more
+                // annoying in practice...
+                data.write_at(position, 0x9f)
+                    .map_err(|_| ClientError::WentAway.fail())?;
+                position += 1;
+            }
+
+            let tid = TaskId(r.tid);
+            let task_name = hubris_task_names::TASK_NAMES
+                .get(tid.index())
+                .copied()
+                .unwrap_or_else(|| {
+                    // This represents an internal error, where we've recorded
+                    // an out-of-range task ID somehow. We still want to get the
+                    // ereport out, so we'll use a recognizable but illegal task
+                    // name to indicate that it's missing.
+                    ringbuf_entry!(Trace::EreportError(
+                        EreportError::TaskIdOutOfRange
+                    ));
+                    "-" // TODO
+                });
+            let generation = tid.generation();
+
+            let entry = (
+                task_name,
+                u8::from(generation),
+                r.timestamp,
+                ByteGather(r.slices.0, r.slices.1),
+            );
+            let mut c =
+                minicbor::encode::write::Cursor::new(&mut self.recv[..]);
+            match minicbor::encode(&entry, &mut c) {
+                Ok(()) => {
+                    let size = c.position();
+                    // If there's no room left for this one in the lease, we're
+                    // done here.
```

Comment:
I was initially concerned whether this leaves room for the final `CBOR_BREAK`.  I now _think_ it does, but adding a comment to make that explicit would be helpful.